### PR TITLE
[SPARK-19800][SS][WIP] Implement one kind of streaming sampling - reservoir sampling

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

This pr adds a special streaming sample operator to support `sample`. It adds a new evolving operator `reservoir`, and introduce an new logical plan `ReservoirSample` and two physical plan `StreamingReservoirSampleExec` and `ReservoirSampleExec`. 

The following cases are supported:

- batch table reservoir sampling
- stream table reservoir sampling with/without aggregation and watermark in Update/Complete output mode

Not supported cases:

- reservoir sampling in Append output mode (No meaning)

Followups:

- move `reservoir` into `sample` operator

## How was this patch tested?

add new unit tests.

### CODE REVIEW

**Comment by srowen** [comment]
Why does this need to be in Spark?

**Comment by uncleGen** [comment]
@srowen There are some unsupported `operator` for Structured Streaming. You can view here: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala#L184
This pr adds support for `sample` operator.

**Comment by uncleGen** [comment]
cc @zsxwing and @tdas

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 1]
```diff
@@ -397,3 +402,110 @@ object StreamingDeduplicateExec {
...
+    outputMode: Option[OutputMode] = None)
+  extends UnaryExecNode with StateStoreWriter with WatermarkSupport {
+
+  override def requiredChildDistribution: Seq[Distribution] =
+  ClusteredDistribution(keyExpressions) :: Nil
+
+  private val enc = Encoders.STRING.asInstanceOf[ExpressionEncoder[String]]
+  private val NUM_RECORDS_IN_PARTITION = enc.toRow("NUM_RECORDS_IN_PARTITION")
+    .asInstanceOf[UnsafeRow]
+
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: `NUM_RECORDS_IN_PARTITION ` calculate the total number of records in current partiton, and update at the end of sample.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 2]
```diff
@@ -397,3 +402,110 @@ object StreamingDeduplicateExec {
...
+      outputMode match {
+        case Some(Complete) =>
+          CompletionIterator[InternalRow, Iterator[InternalRow]](
+            store.iterator().filter(kv => {
+              !kv._1.asInstanceOf[UnsafeRow].equals(NUM_RECORDS_IN_PARTITION)
+            }).map(kv => {
+              UnsafeProjection.create(withSumFieldTypes).apply(InternalRow.fromSeq(
+                new JoinedRow(kv._2, numRecordsTillNow)
+                  .toSeq(withSumFieldTypes)))
+            }), {})
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: Here, we transfer the row to (row, numRecordsTillNow), and `numRecordsTillNow` is used to calculate the weight of item.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 3]
```diff
@@ -397,3 +402,110 @@ object StreamingDeduplicateExec {
...
+        case Some(Update) =>
+          CompletionIterator[InternalRow, Iterator[InternalRow]](
+            store.updates()
+              .filter(update => !update.key.equals(NUM_RECORDS_IN_PARTITION))
+              .map(update => {
+                UnsafeProjection.create(withSumFieldTypes).apply(InternalRow.fromSeq(
+                    new JoinedRow(update.value, numRecordsTillNow)
+                      .toSeq(withSumFieldTypes)))
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: same

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 4]
```diff
@@ -397,3 +402,110 @@ object StreamingDeduplicateExec {
...
+      }
+    }.repartition(1).mapPartitions(it => {
+      SamplingUtils.reservoirSampleWithWeight(
+        it.map(item => (item, item.getLong(keyExpressions.size))), k)
+        .map(row =>
+          UnsafeProjection.create(fieldTypes)
+            .apply(InternalRow.fromSeq(row.toSeq(fieldTypes)))
+        ).iterator
+    })
+  }
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: here, we do once global weight reservoir sampling.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 5]
```diff
@@ -397,3 +402,110 @@ object StreamingDeduplicateExec {
...
+          store.put(enc.toRow(numSamples.toString).asInstanceOf[UnsafeRow],
+            r.asInstanceOf[UnsafeRow])
+        } else {
+          val randomIdx = (rand.nextDouble() * (numRecordsInPart + count)).toLong
+          if (randomIdx <= k) {
+            val replacementIdx = enc.toRow(randomIdx.toString).asInstanceOf[UnsafeRow]
+            store.put(replacementIdx, r.asInstanceOf[UnsafeRow])
+          }
+        }
+      }
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: In partiton, we just need to do once normal (without weight) reservoir sampling.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/ReservoirSampleSuit.scala [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  test("batch reservoir sample before aggregation") {
+    val df = spark.createDataset((1 to 10).map(e => (e, s"val_$e")))
+      .select($"_1" as "key", $"_2" as "value")
+      .reservoir(3)
+      .groupBy("value").count()
+    assert(df.count() == 3, "")
+  }
+}
```
Comment (uncleGen) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/ReservoirSampleSuit.scala]: new unit test needs to be improved.

**Comment by uncleGen** [comment]
ping @tdas and @zsxwing

**Comment by jiangxb1987** [comment]
Is anyone still working on this or this could be closed? @uncleGen @zsxwing

**Pull request closed without merging**